homeassistant/components/backup/config.py

```diff
@@ -271,16 +338,81 @@ def _delete_filter(
                 """Return backups older than days to delete."""
                 # we need to check here since we await before
                 # this filter is applied
-                if self.days is None:
+                agents_retention = {
+                    agent_id: agent_config.retention
+                    for agent_id, agent_config in manager.config.data.agents.items()
+                }
+                has_agents_retention = any(
+                    agent_retention for agent_retention in agents_retention.values()
+                )
+                has_agents_retention_days = any(
+                    agent_retention and agent_retention.days is not None
+                    for agent_retention in agents_retention.values()
+                )
+                if (global_days := self.days) is None and not has_agents_retention_days:
+                    # No global retention days and no agent retention days
                     return {}
+
                 now = dt_util.utcnow()
-                return {
-                    backup_id: backup
-                    for backup_id, backup in backups.items()
-                    if dt_util.parse_datetime(backup.date, raise_on_error=True)
-                    + timedelta(days=self.days)
-                    < now
-                }
+                if global_days is not None and not has_agents_retention:
+                    # Return early to avoid the longer filtering below.
+                    return {
+                        backup_id: backup
+                        for backup_id, backup in backups.items()
+                        if dt_util.parse_datetime(backup.date, raise_on_error=True)
+                        + timedelta(days=global_days)
+                        < now
+                    }
+
+                # If there are any agent retention settings, we need to check
+                # the retention settings, for every backup and agent combination.
+
+                backups_to_delete = {}
+
+                for backup_id, backup in backups.items():
+                    backup_date = dt_util.parse_datetime(
+                        backup.date, raise_on_error=True
+                    )
+                    agents_for_backup_to_delete = set(backup.agents)
```

Comment:
I think this variable name is a bit hard to understand.
The set is updated so at the end of the iteration it has the agents from which the backup should be deleted? Maybe name it `delete_from_agents` or something?